Java 15 Use Unsafe.defineAnonyousClass() to define Hidden classes.

hangshao0 opened the pull request:
Add implementation of new Hidden class APIs to unblock the JDK15 builds.

issue #9328

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

hangshao0 commented:
This PR is based on #9292. The last commit is my change. Other changes have already been review under #9292.

hangshao0 commented:
Reviewer: @pshipton

hangshao0 commented:
More change is required to resolve #9328. The change here is to unblock building of the latest jdk15.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -369,7 +369,7 @@ private static boolean isSamePackage(Class<?> a, Class<?> b){
 		
 		void checkAccess(MethodHandle handle, boolean skipAccessCheckPara) throws IllegalAccessException {
-			if (INTERNAL_PRIVILEGED == accessMode) {
+			if (INTERNAL_PRIVILEGED == accessMode || FULL_ACCESS_MASK == accessMode) {
```
Please add extra parentheses here and in similar tests on lines 391 amd 405.

hangshao0 replied:
Fixed.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2256,17 +2256,38 @@ public boolean hasFullPrivilegeAccess() {
 
 		public Lookup defineHiddenClass(byte[] bytes, boolean initOption, ClassOption... classOptions) {
-			return null;
+			ClassReader cr = null;
```
Please remove the unnecessary initialization.

hangshao0 replied:
Fixed.

keithc-ca commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2256,17 +2256,38 @@ public boolean hasFullPrivilegeAccess() {
-				return null;
+			private byte[] classBytes;
+			private String className;
+			private Lookup lookup;
```
The class and all its fields should be `final`.

hangshao0 replied:
Fixed.

pshipton commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2255,18 +2261,39 @@ public boolean hasFullPrivilegeAccess() {
+			try {
+				cr = new ClassReader(bytes);
+			} catch (ArrayIndexOutOfBoundsException e) {
+				/*[MSG "K065Y2", "The class byte array is corrupted"]*/
```
This should be defined in jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties.

hangshao0 replied:
It is already there: 
https://github.com/eclipse/openj9/blob/master/jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties#L407

pshipton commented:
jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

pshipton commented:
Set as WIP since when this is merged, the jdk openj9-staging branch also needs to be promoted. We'll need to run a full acceptance build first.

keithc-ca commented:
> jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

That didn't work: it built openj9 instead of openj9-staging.

AdamBrousseau commented:
The trigger is open enough that any type of text/code/quoting `jenkins` followed by `compile` or `test` will trigger a build.

keithc-ca commented:
This error is unexpected:
```
[2020-05-07T18:42:21.333Z] /home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/openj9/debugtools/DDR_VM/src/com/ibm/j9ddr/view/dtfj/image/J9DDRImage.java:336: error: method does not override or implement a method from a supertype
[2020-05-07T18:42:21.333Z] 	@Override
[2020-05-07T18:42:21.333Z] 	^
```
I'll investigate.

pshipton commented:
The build error is the following which I think we've seen before. @keithc-ca are you going to fix this one?
```
/home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/openj9/debugtools/DDR_VM/src/com/ibm/j9ddr/view/dtfj/image/J9DDRImage.java:336: error: method does not override or implement a method from a supertype
[2020-05-07T18:42:21.333Z] 	@Override
```

keithc-ca commented:
It's unclear how that could happen. The `DDR.gmk` makefile is being executed for the `openj9.dtfj-ddr-jar` target which depends upon `openj9.dtfj-ddr-gen` which in turn depends on `openj9.dtfj-java`: when `openj9.dtfj-java` has completed we should expect the class files for that module to be available.

The previous problem was due improper javac options, but that has been fixed: see ibmruntimes/openj9-openjdk-jdk#199.

keithc-ca commented:
Jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

keithc-ca commented:
Jenkins test sanity zlinux jdknext depends keithc-ca/openj9-openjdk#makefiles

pshipton commented:
Jenkins test sanity zlinux jdknext depends keithc-ca/openj9-openjdk-jdk#makefiles

keithc-ca commented:
That incantation doesn't appear to work: I think it needs to reference a pull request (instead of a branch in some random repository).

pshipton commented:
> That incantation doesn't appear to work: I think it needs to reference a pull request (instead of a branch in some random repository).

The one you started had the wrong repo name, openj9-openjdk instead of openj9-openjdk-jdk

pshipton commented:
The machines are offline for a short while for sanitizing, the job is waiting on a machine. We'll see if it works in a bit.

keithc-ca commented:
> The one you started had the wrong repo name, openj9-openjdk instead of openj9-openjdk-jdk

No, that is the name I use for that repo (I omit the `-jdk` segment).

pshipton commented:
Jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

pshipton commented:
Jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

AdamBrousseau commented:
The depends, on Keith's repo, doesn't work because at the moment, we only have the ability to fetch refs from the same repo as the one that is cloned by default. The exception to this is omr, where we have the option to fetch from the main omr repo instead of the openj9-omr repo. Functionality to fetch from any fork could be added but there's some added risk involved. I would suggest opening an issue if it's a feature we want.

DanHeidinga commented:
I'm going to be slow to review changes this week due to some talks I'm giving.

pshipton commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2255,18 +2261,39 @@ public boolean hasFullPrivilegeAccess() {
+			Class<?> defineClass(boolean initOption) {
+				JavaLangAccess jlAccess = SharedSecrets.getJavaLangAccess();
+				Class<?> lookupClass = lookup.lookupClass();
+				return jlAccess.defineClass(lookupClass.getClassLoader(), lookupClass, className, classBytes, jlAccess.protectionDomain(lookupClass), initOption, 0, null);
```
@hangshao0 pls wrap this in an AccessController.doPrivileged() to fix some of the test failures.

hangshao0 replied:
Fixed.

pshipton commented:
jenkins test sanity zlinux jdk11

pshipton commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -369,7 +369,9 @@ private static boolean isSamePackage(Class<?> a, Class<?> b){
 		void checkAccess(MethodHandle handle, boolean skipAccessCheckPara) throws IllegalAccessException {
-			if (INTERNAL_PRIVILEGED == accessMode) {
+			if ((INTERNAL_PRIVILEGED == accessMode)
+				|| (FULL_ACCESS_MASK == accessMode)
```
Seems adding this is causing a number of test suite failures, even in jdk11.
https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_sanity.functional_s390x_linux_Personal/413

hangshao0 replied:
I added this line because OpenJ9 is throwing IllegalAccessException with message "K0587". The new Java 15 code from InnerClassLambdaMetafactory is passing in accessMode FULL_ACCESS_MASK. It is calling from findConstructor().

hangshao0 replied:
OK, I seems to know what is going on. definingClass should be made nestmate of accessClass which is missing here.

hangshao0 replied:
I have removed the checks for FULL_ACCESS_MASK. 

I don't find an existing APIs (Java or native) that could let me make a class to be nestmate of another class. So I have changed the native implementation of getNestHost() to return the nesthost of the loopup class for hidden classes (which is defined by Unsafe.defineAnonyousClass()). 

I am able to build Java 15 with the current change.

pshipton commented:
jenkins test sanity zlinux jdk11

pshipton commented:
Jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

pshipton commented:
Windows is broken atm, will have to test it tomorrow after it's fixed.

Jenkins test sanity aix,plinux,zlinux,xlinux,xlinuxcm,xlinuxxl,osx jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

pshipton commented:
Seems there is a new problem since the last update of jdknext material.
``
01:14:29  Building target 'all' in configuration 'linux-s390x-server-release'
01:14:30  Compiling /home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/make/langtools/tools/compileproperties/CompileProperties.java /home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/make/langtools/tools/propertiesparser/PropertiesParser.java /home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/make/langtools/tools/propertiesparser/gen/ClassGenerator.java /home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/make/langtools/tools/propertiesparser/parser/Message.java /home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/make/langtools/tools/propertiesparser/parser/MessageFile.java /home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/make/langtools/tools/propertiesparser/parser/MessageInfo.java /home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/make/langtools/tools/propertiesparser/parser/MessageLine.java /home/jenkins/workspace/Build_JDKnext_s390x_linux_Personal/make/langtools/tools/propertiesparser/parser/MessageType.java files for BUILD_TOOLS_LANGTOOLS
01:14:30  gmake[2]: *** No rule to make target 'test-image-failure-handler', needed by 'test-image'.  Stop.
```

keithc-ca commented:
> Seems there is a new problem since the last update of jdknext material.
 
Significant changes were made to `make/Main.gmk` involving the use of a new macro, `SetupTargetBody`. Hopefully it won't be too difficult to adapt.

keithc-ca commented:
ibmruntimes/openj9-openjdk-jdk#207 should fix the `test-image-failure-handler` problem.

keithc-ca commented:
Jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#207

pshipton commented:
Jenkins test sanity aix,plinux,zlinux,xlinux,xlinuxcm,xlinuxxl,osx,windows jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

keithc-ca commented on runtime/jcl/common/java_lang_Class.cpp:
```diff
@@ -1805,13 +1805,26 @@ Java_java_lang_Class_getNestHostImpl(JNIEnv *env, jobject recv)
-		if (J9_VISIBILITY_ALLOWED == vmFuncs->loadAndVerifyNestHost(currentThread, clazz, J9_LOOK_NO_THROW)) {
-			nestHost = clazz->nestHost;
+		J9Class *clazzToUse = clazz;
+		J9JavaVM *vm = currentThread->javaVM;
```
This local could be pushed before, and reused in, the declaration of `vmFuncs`.

hangshao0 replied:
Changed the code to use `#if JAVA_SPEC_VERSION >= 15`. There is no need to have this local variable `vm` anymore.

keithc-ca commented on runtime/jcl/common/java_lang_Class.cpp:
```diff
@@ -1805,13 +1805,26 @@ Java_java_lang_Class_getNestHostImpl(JNIEnv *env, jobject recv)
-			nestHost = clazz->nestHost;
+		J9Class *clazzToUse = clazz;
+		J9JavaVM *vm = currentThread->javaVM;
+		if (J2SE_VERSION(vm) >= J2SE_V15) {
```
Why not `#if JAVA_SPEC_VERSION >= 15` instead of runtime check?

hangshao0 replied:
I've changed the code to use `#if JAVA_SPEC_VERSION >= 15`

keithc-ca commented:
Jenkins test sanity aix,plinux,xlinux,xlinuxcm,xlinuxxl,osx,win jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

(fixes platform name windows -> win)

keithc-ca commented:
Changes required: https://github.com/eclipse/openj9/pull/9292#discussion_r424610735

pshipton commented:
@hangshao0 I've updated https://github.com/eclipse/openj9/pull/9292 to fix the Windows compilation problem, can you please pick up this change.

The last PR build is
https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/3553/
The zlinux one https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/3551/ passed.

hangshao0 commented:
> I've updated #9292 to fix the Windows compilation problem, can you please pick up this change.

I've added the change requested by https://github.com/eclipse/openj9/pull/9292#discussion_r424610735 to my latest commit.

keithc-ca commented:
Jenkins test sanity aix,plinux,zlinux,xlinux,xlinuxcm,xlinuxxl,osx,win jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

keithc-ca commented:
@hangshao0 Please remove the 'WIP' prefix if you agree this is ready to be merged.

pshipton commented:
I added the WIP. It can be removed as long as we're ready to merge this and update jdknext to the tested openj9-staging level.

keithc-ca commented:
The failed test (cmdLineTester_callsitedbgddrext_openj9_0) is a known issue (#90).

pshipton commented:
Testing lgtm. We can ignore the xlinuxcm failure related to the unreadable core file.

DanHeidinga commented on runtime/j9vm/java11vmi.c:
```diff
@@ -710,18 +730,80 @@ allowReadAccessToModule(J9VMThread * currentThread, J9Module * fromModule, J9Mod
+					oom = TRUE;
+					break;
+				}
+				if (JNI_FALSE == isCopy) {
```
`isCopy` is an output parameter.  This code needs to handle the case where the string isn't a copy

hangshao0 replied:
Updated to use `copyStringToUTF8Helper()`. `isCopy` is removed.

DanHeidinga commented on runtime/j9vm/java11vmi.c:
```diff
@@ -710,18 +730,80 @@ allowReadAccessToModule(J9VMThread * currentThread, J9Module * fromModule, J9Mod
+					Assert_SC_unreachable();
+					goto cleanup;
+				}
+				convertPackageName((char*)packageName);
```
It would be more efficient to get vm access earlier and do this with VM calls rather than JNI.  It would also remove the need to assert on the `isCopy` above by using the `copyStringToUTF8Helper` helper

hangshao0 replied:
Updated to use copyStringToUTF8Helper().

DanHeidinga commented on runtime/j9vm/java11vmi.c:
```diff
@@ -851,6 +933,29 @@ JVM_DefineModule(JNIEnv * env, jobject module, jboolean isOpen, jstring version,
+				break;
+			}
+			(*env)->ReleaseStringUTFChars(env, packageObj, packageName);
+			(*env)->DeleteLocalRef(env, packageObj);
```
same here - keep vm access and use the vm helpers rather than JNI

hangshao0 replied:
Updated to use copyStringToUTF8Helper().

DanHeidinga commented on runtime/j9vm/java11vmi.c:
```diff
@@ -866,19 +971,46 @@ JVM_DefineModule(JNIEnv * env, jobject module, jboolean isOpen, jstring version,
+#if JAVA_SPEC_VERSION >= 15
+	jboolean isCopy = JNI_FALSE;
+	const char* package = (NULL == packageObj) ? NULL : (*env)->GetStringUTFChars(env, packageObj, &isCopy);
+	if ((NULL != package) && (JNI_FALSE == isCopy)) {
```
This should also move after getting vm access and use the `copyStringToUTF8Helper` helper.  Note, there's an `J9_STR_XLAT` flag that does the `.` -> `/` conversion

hangshao0 replied:
Updated to use copyStringToUTF8Helper().

DanHeidinga commented on runtime/j9vm/java11vmi.c:
```diff
@@ -913,19 +1056,46 @@ JVM_AddModuleExports(JNIEnv * env, jobject fromModule, const char *package, jobj
+#if JAVA_SPEC_VERSION >= 15
+	jboolean isCopy = JNI_FALSE;
+	const char* package = (NULL == packageObj) ? NULL : (*env)->GetStringUTFChars(env, packageObj, &isCopy);
+	if ((NULL != package) && (JNI_FALSE == isCopy)) {
```
same here as well

hangshao0 replied:
Updated to use copyStringToUTF8Helper().

DanHeidinga commented on runtime/j9vm/java11vmi.c:
```diff
@@ -1142,19 +1323,46 @@ JVM_AddModulePackage(JNIEnv * env, jobject module, const char *package)
 	J9InternalVMFunctions const * const vmFuncs = vm->internalVMFunctions;
+#if JAVA_SPEC_VERSION >= 15
+	jboolean isCopy = JNI_FALSE;
+	const char* package = (NULL == packageObj) ? NULL : (*env)->GetStringUTFChars(env, packageObj, &isCopy);
```
and here

hangshao0 replied:
Updated to use copyStringToUTF8Helper().

hangshao0 commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2245,7 +2246,57 @@ public boolean hasFullPrivilegeAccess() {
+		/*[IF Java15]*/
+		// TODO: implement support for hidden classes.
+
+		public enum ClassOption {
```
Javadoc added.

hangshao0 commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -2245,7 +2246,57 @@ public boolean hasFullPrivilegeAccess() {
+			}
+		}
+
+		public Lookup defineHiddenClass(byte[] bytes, boolean initOption, ClassOption... classOptions) {
```
Javadoc added. 

> `classOptions` will be used later in the complete implementation?

Yes.

keithc-ca commented on runtime/j9vm/java11vmi.c:
```diff
@@ -710,18 +711,79 @@ allowReadAccessToModule(J9VMThread * currentThread, J9Module * fromModule, J9Mod
+		jsize pkgIndex = 0;
+		memset((void *)packages, 0, packagesNumBytes);
+		for (pkgIndex = 0; pkgIndex < numPackages; pkgIndex++) {
+			jobject packageObj = (*env)->GetObjectArrayElement(env, packageArray, pkgIndex);
+			if ((*env)->ExceptionCheck(env)) {
+				(*env)->DeleteLocalRef(env, packageObj);
```
See https://github.com/eclipse/openj9/pull/9292#discussion_r412239235: using JNI after calling `internalEnterVMFromJNI()` is not legal.

hangshao0 replied:
Removed those JNI calls.

DanHeidinga commented on runtime/j9vm/java11vmi.c:
```diff
@@ -710,18 +711,72 @@ allowReadAccessToModule(J9VMThread * currentThread, J9Module * fromModule, J9Mod
+		}
+	}
+
+	if (NULL == packageArray) {
```
Can this check be moved earlier?  It seems like we do a `!= NULL` check earlier that can be commoned with this

hangshao0 replied:
Yes, it can be moved earlier. Done.

DanHeidinga commented on runtime/j9vm/java11vmi.c:
```diff
@@ -710,18 +711,72 @@ allowReadAccessToModule(J9VMThread * currentThread, J9Module * fromModule, J9Mod
+				vmFuncs->copyStringToUTF8Helper(currentThread, stringObject, J9_STR_NULL_TERMINATE_RESULT | J9_STR_XLAT , 0, J9VMJAVALANGSTRING_LENGTH(currentThread, stringObject), packageName, utfLength);
+				packages[pkgIndex] = packageName;
+			} else {
+				nullPackage = TRUE;
```
Is the `nullPackage` variable needed?  Can the exception setting code be moved here and then the code could goto directly the `cleanup` label?

It means there's one less piece of state to keep straight

hangshao0 replied:
Removed variable `nullPackage` and moved the exception setting here.

DanHeidinga commented on runtime/j9vm/java11vmi.c:
```diff
@@ -866,19 +932,46 @@ JVM_DefineModule(JNIEnv * env, jobject module, jboolean isOpen, jstring version,
+	if (NULL != packageObj) {
+		j9object_t stringObject = J9_JNI_UNWRAP_REFERENCE(packageObj);
+		UDATA utfLength = vmFuncs->getStringUTF8Length(currentThread, stringObject) + 1;
+		char* packageName = (char *)j9mem_allocate_memory(utfLength, OMRMEM_CATEGORY_VM);
```
similar potential optimization here by keeping a small stack allocated buffer to avoid the malloc

hangshao0 replied:
Will open a separate issue once this piece of code is merged.

hangshao0 commented on runtime/j9vm/java11vmi.c:
```diff
@@ -710,18 +711,72 @@ allowReadAccessToModule(J9VMThread * currentThread, J9Module * fromModule, J9Mod
+	if ((*env)->ExceptionCheck(env)) {
+		return module;
+	}
+#endif /* JAVA_SPEC_VERSION >= 15 */
```
Fixed.

hangshao0 commented on runtime/j9vm/java11vmi.c:
```diff
@@ -710,18 +711,72 @@ allowReadAccessToModule(J9VMThread * currentThread, J9Module * fromModule, J9Mod
+	jsize numPackages = (NULL == packageArray) ? 0 : (*env)->GetArrayLength(env, packageArray);
+	PORT_ACCESS_FROM_ENV(env);
+	UDATA packagesNumBytes = sizeof(char*) * numPackages;
+	const char** packages = NULL;
```
Yes, I will open a separate issue once this piece of code is merged.

hangshao0 commented on runtime/j9vm/java11vmi.c:
```diff
@@ -710,18 +711,72 @@ allowReadAccessToModule(J9VMThread * currentThread, J9Module * fromModule, J9Mod
 #endif /* defined(CALL_BUNDLED_FUNCTIONS_DIRECTLY) */
+
+#if JAVA_SPEC_VERSION >= 15
+	packages = (const char**)j9mem_allocate_memory(packagesNumBytes, OMRMEM_CATEGORY_VM);
```
Fixed.

DanHeidinga commented:
Jenkins test sanity aix,zlinux,win jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

DanHeidinga commented:
Running testing on a limited subset of platforms to catch any compiler errors given previous test runs were largely clean

pshipton commented:
@DanHeidinga note that ibmruntimes/openj9-openjdk-jdk#openj9-staging is a moving target, it's typically updated every night. We should do a full run of the platforms*, and then merge this and promote openj9-staging before it gets updated in the nightly merge from OpenJDK. Otherwise, the tested openj9-staging sha can be promoted manually, which hopefully is just a push if no additional jdknext PRs get merged in the meantime.

* the subset of platforms tested by the acceptance build are:
ppc64_aix,x86-64_linux,x86-64_linux_xl,x86-64_linux_cm,ppc64le_linux,s390x_linux,x86-64_windows,x86-64_mac

pshipton commented:
The AIX failure is 
```
14:23:59  /home/jenkins/workspace/Build_JDKnext_ppc64_aix_Personal/build/aix-ppc64-server-release/support/j9jcl_sources/java.base/share/classes/java/lang/Access.java:61: error: Access is not abstract and does not override abstract method stringConcatInitialCoder() in JavaLangAccess
14:23:59  final class Access implements JavaLangAccess {
14:23:59        ^
```

pshipton commented:
@hangshao0 can you please stub the new method so we can re-test.

hangshao0 commented:
> can you please stub the new method so we can re-test.

Done.

keithc-ca commented on runtime/j9vm/java11vmi.c:
```diff
@@ -843,6 +892,17 @@ JVM_DefineModule(JNIEnv * env, jobject module, jboolean isOpen, jstring version,
+		j9mem_free_memory((void *)packages);		
+	}
+#endif /* JAVA_SPEC_VERSION >= 15 */
 done:
```
The `goto` on line 745 should target `cleanup` to avoid leaking `packages`.
That would be the last use of `done`; perhaps `cleanup` should just become `done`.

hangshao0 replied:
Fixed.

keithc-ca commented:
Jenkins test sanity aix,osx,plinux,win,xlinux,xlinuxcm,xlinuxxl,zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

pshipton commented:
Jenkins test sanity aix,zlinux,win,osx jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

pshipton commented:
sorry @keithc-ca, for some reason I didn't your PR build before I started mine. I stopped mine, your's is still running https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/3621/

pshipton commented:
@hangshao0 pls fix the commit title for "Add Javadoc to ClassOption and defineHiddenClass()". This title is misleading since the commit doesn't only add javadoc changes.

hangshao0 commented:
Add more info into the title.

keithc-ca commented:
@hangshao0 Can you please update the commit comment again?
The first line should be no more than 70 characters and the second line should be blank:
see https://github.com/eclipse/openj9/blob/master/CONTRIBUTING.md.

hangshao0 commented:
> Can you please update the commit comment again?

Done.

keithc-ca commented:
Jenkins test sanity aix,osx,win,win32,zlinux jdk8

keithc-ca commented:
Jenkins test sanity aix,osx,plinux,win jdk11
